doc/man7/openssl-threads.pod

```diff
@@ -0,0 +1,100 @@
+=pod
+
+=head1 NAME
+
+openssl-threads - Overview of thread safety in OpenSSL
+
+=head1 DESCRIPTION
+
+In this man page, we use the term B<thread-safe> to indicate that an
+object can be used by multiple threads at the same time.
+
+OpenSSL can be built with or without threads support. The most important
+use of this support is so that OpenSSL itself can use a single consistent
+API, as shown in L<CRYPTO_THREAD_run_once(3)/EXAMPLES>.
+Multi-platform applications can also use this API.
+
+In particular, being configured for threads support does not imply that
+all OpenSSL objects are thread-safe.
+To repeat: I<most objects are not safe for simultaneous use>.
+Exceptions to this should be documented on the specific manual pages, and
+some general high-level guidance is given here.
+
+One major use of the OpenSSL thread API is to implement reference counting.
+Many objects within OpenSSL are reference-counted, so resources are not
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is often an I<object>_up_ref() API that can be used to increase the
+reference.
+Failure to match I<object>_up_ref() calls with the right number of
+I<object>_free() calls is a common source of memory leaks when a program
+exits.
+
+The various C<set1> and C<get1> functions associated with an object are
+used to set or get attributes of an object; they increase the reference
+count of the attribute, if the attribute is an object that has them.
+The C<set0> and C<get0> functions are similar, but do not increase the
+reference count, passing "ownership" appropriately.
```

Comment:
I think this is not entirely clear. A "set0" passes ownership from the caller to the object. However with a "get0" ownership remains with the object.

With a "set1" ownership remains with the caller, and with a "get1" ownership is passed from the object to the caller.

Reply:
Thanks for that input, i used it to rewrite like this:
```
Many objects have set and get API's to set attributes in the object. A C<set0> passes ownership
from the caller to the object and a  C<get0> returns a pointer but the attribute ownership remains
with the object. A C<set1> or C<get1> function does not change the ownership, and it is up to the
caller to free the object or resource. Functions with a simple C<set> or C<get> are historical; and
the documentation must state how the references are handled. Get methods are often thread-safe
as long as the ownership requirements are met. Set methods are generally not thread-safe as discussed below.
```

Reply:
I don't know if it's always the case, but a get1 and set1 increases the reference count, and so a for get1 the application should free it, for a set1 it will be freed as part of freeing the struct. I guess an other implementation of a get1 and set1 can be that it duplicates it, but it should always free it. So either the ownership is moved or shared.

Reply:
Sorry, I don't understand what you're trying to say.  That's my fault.  Can you propose specific wording changes to the text quoted in the comment above?

Reply:
You say `A C<set1> or C<get1> function does not change the ownership`, but it does get1 does.

Reply:
Thanks!  How about "set1 transfers ownership of the attribute to the object and get1 returns a copy" ?

Reply:
I guess I edited my previous reply before you saw it.

Reply:
The `1` functions always make a copy, which can be implemented as a reference count increment. So on get, it becomes a (shared) owner, on set1 it stays a (shared) owner.

Reply:
> The `1` functions always make a copy, which can be implemented as a reference count increment. So on get, it becomes a (shared) owner, on set1 it stays a (shared) owner.

Manipulating the refcount isn't a copy tho.  It's `foo->p = p` versus `foo->p = strdup(p);`

But anyway how about this:
```
Many objects have set and get API's to set attributes in the object.
A C<set0> passes ownership from the caller to the object and a
C<get0> returns a pointer but the attribute ownership
remains with the object and a copy is returned.
A C<set1> or C<get1> function does not change the ownership, but instead
updates the attribute's reference count so that the object is shared 
between the caller and the object; the caller must free the attribute when
finished.
Functions with a simple C<set> or C<get> are historical; and the documentation
must state how the references are handled.
Get methods are often thread-safe as long as the ownership requirements are
met and shared objects are not modified.
Set methods, or modifying shared objects, are generally not thread-safe 
as discussed below.
```

Reply:
`s/remains with the object and a copy is returned/remains with the object and just a reference to it is returned`

Are there any cases where a full dup operation is done on set1/get1? Not that it matters much as it is always better to treat them as shared reference just with an increased refcount.

Reply:
> Are there any cases where a full dup operation is done on set1/get1?

strdup kinds of things.  but sure, your reasoning makes sense; fixup commit pushed.